### PR TITLE
[Shader Management] Foundry Additional Template Settings - Graphics Repo

### PR DESCRIPTION
**Pull request by cdxntchou**
### Purpose of this PR

This PR has the corresponding changes for the following trunk PRs:
https://github.cds.internal.unity3d.com/unity/unity/pull/7965
https://github.cds.internal.unity3d.com/unity/unity/pull/10598

* Shader Dependency declarations
* UsePass declarations
* CustomEditor declarations
* Debug flag on Template pass
* Shader Fallback declarations
* Shader Function Include support

Prototype update PR: https://github.cds.internal.unity3d.com/joshua-davis/BlockSurfaceShaders/pull/2

---
### Testing status
Added tests to the ShaderFoundry, testing:
* Shader Dependency declarations show up when appropriate
* Custom Editor declarations show up when appropriate
* Shader Fallback declarations show up when appropriate
* Include declarations show up when appropriate (both in Block includes and Block Function includes)
---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.shadergraph/Editor/ShaderFoundry/Provider/LegacyTemplateProvider.cs [hunk 1]
```diff
@@ -87,6 +88,24 @@ Template BuildTemplate(SubShaderDescriptor subShaderDescriptor, int subShaderInd
...
+
+            if (subShaderDescriptor.shaderDependencies != null)
+                foreach (var dependency in subShaderDescriptor.shaderDependencies)
+                    builder.AddShaderDependency(dependency.dependencyName, dependency.shaderName);
+
+            if (subShaderDescriptor.shaderCustomEditors != null)
+                foreach (var customEditor in subShaderDescriptor.shaderCustomEditors)
+                    builder.AddShaderCustomEditor(customEditor.shaderGUI, customEditor.renderPipelineAssetType);
+
+            // TODO: we might just be able to get rid of subShaderDescriptor.shaderCustomEditor,
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/ShaderFoundry/Provider/LegacyTemplateProvider.cs:102]: usePassList should also be extracted in here.

**Review by joshua-davis — CHANGES_REQUESTED** [review]

File: com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs [hunk 2]
```diff
@@ -1,27 +1,88 @@
...
+                    else
+                    {
+                        if (m_FallbackShader != templateInstance.Template.ShaderFallback)
+                        {
+                            // ERROR : conflicting shader fallbacks defined...
+                        }
+                    }
+                }
+
+                foreach (var customEditor in templateInstance.Template.ShaderCustomEditors)
```
[the remark this answers is not in the document]
Comment (cdxntchou) [reply com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs:72]: They are actually different checks.  But I can combine them here instead of splitting them in two places.

File: com.unity.shadergraph/Editor/ShaderFoundry/Provider/LegacyTemplateProvider.cs [hunk 3]
```diff
@@ -87,6 +88,24 @@ Template BuildTemplate(SubShaderDescriptor subShaderDescriptor, int subShaderInd
                 builder.AddPass(passBuilder.Build());
             }
 
+            builder.AdditionalShaderID = subShaderDescriptor.additionalShaderID;
```
[the remark this answers is not in the document]
Comment (cdxntchou) [reply com.unity.shadergraph/Editor/ShaderFoundry/Provider/LegacyTemplateProvider.cs:101]: yup, this was missing.

File: com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs [hunk 4]
```diff
@@ -41,8 +101,21 @@ public Builder(ShaderContainer container, string name, string additionalShaderID
 
             public ShaderInstance Build()
             {
-                return new ShaderInstance(Name, AdditionalShaderID, TemplateInstances, FallbackShader);
+                return new ShaderInstance(Name, AdditionalShaderID, TemplateInstances);
             }
         }
     }
+
+    // we must declare this sorter class, instead of making ShaderDependency implement IComparable, because
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs:109]: I'd investigate this more, maybe a clean rebuild?

File: com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs [hunk 5]
```diff
@@ -8,20 +10,79 @@ namespace UnityEditor.ShaderFoundry
...
+            {
+                if (!string.IsNullOrEmpty(templateInstance.Template.ShaderFallback))
+                {
+                    if (m_FallbackShader == null)
+                        m_FallbackShader = templateInstance.Template.ShaderFallback;
+                    else
+                    {
+                        if (m_FallbackShader != templateInstance.Template.ShaderFallback)
+                        {
+                            // TODO @ SHADERS ERROR - conflicting shader fallbacks defined...
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs:46]: You may consider using `ErrorHandling.ReportError` or adding to this so that we have a unified place to update later instead of a bunch of separate ones.

File: com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderGenerator.cs [hunk 6]
```diff
@@ -102,5 +103,25 @@ static void GenerateSubShaders(ShaderBuilder builder, ShaderContainer container,
...
+        static void GenerateDependencies(ShaderBuilder builder, ShaderContainer container, ShaderInstance shaderInst)
+        {
+            foreach (var dependency in shaderInst.Dependencies)
+                builder.AppendLine($"Dependency \"{dependency.DependencyName}\" = \"{dependency.ShaderName}\"");
+        }
+
+        static void GenerateCustomEditors(ShaderBuilder builder, ShaderContainer container, ShaderInstance shaderInst)
+        {
+            foreach (var customEditor in shaderInst.CustomEditors)
+                builder.AppendLine($"CustomEditorForRenderPipeline \"{customEditor.CustomEditorClassName}\" \"{customEditor.RenderPipelineAssetClassName}\"");
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderGenerator.cs:116]: Does this handle the case that I think was still in built-in where it did the old method (non render pipeline specific)?

Comment (cdxntchou) [reply com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderGenerator.cs:116]: The built-in editor and user-specified override is supported via this path.
But I'm thinking the "ShaderGraph default editor" is not...   you would only get that currently if you inspect a material while using a render pipeline that is not supported by the graph.   But we might want to add support for injecting the ShaderGraph default to keep it equivalent.

File: com.unity.shadergraph/Editor/ShaderFoundry/Linker/LegacyTemplateLinker.cs [hunk 7]
```diff
@@ -112,6 +122,9 @@ void AddBlockInstanceProperties(BlockInstance blockInstance)
                 if (!blockInstance.IsValid)
                     return;
 
+                // this looks through block inputs, looking for any with a [Property] attribute, and assumes it is a property.
+                // this may need to be changed for VFX support, that replaces per-instance properties with override values
+                // Add will only add if it isn't already added.
```
Comment (jessebarker) [inline com.unity.shadergraph/Editor/ShaderFoundry/Linker/LegacyTemplateLinker.cs:127]: Yay comments!

File: com.unity.shadergraph/Editor/ShaderFoundry/Provider/LegacyTemplateProvider.cs [hunk 8]
```diff
@@ -86,17 +87,49 @@ void AddTemplateTag(string tagName, string tagValue)
...
+            if (!string.IsNullOrWhiteSpace(subShaderDescriptor.shaderCustomEditor))
+            {
+                shaderGUI = subShaderDescriptor.shaderCustomEditor;
+                rpAssetType = null;
+            }
 
-                builder.AddPass(passBuilder.Build());
+            if (subShaderDescriptor.shaderCustomEditors != null)
+            {
+                foreach (var customEditor in subShaderDescriptor.shaderCustomEditors)
```
Comment (jessebarker) [inline com.unity.shadergraph/Editor/ShaderFoundry/Provider/LegacyTemplateProvider.cs:123]: Last one in the list wins?  Is that the behavior we want?  (not saying it's wrong, just asking)

Comment (cdxntchou) [reply com.unity.shadergraph/Editor/ShaderFoundry/Provider/LegacyTemplateProvider.cs:123]: Yeah.  None of the existing legacy targets are providing more than one anyways...   :)

File: TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/FeatureTests/FallbackTests.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+using System.Linq;
+using NUnit.Framework;
+using UnityEditor.ShaderGraph;
+
+namespace UnityEditor.ShaderFoundry.UnitTests
+{
+    class FallbackTests // : BlockTestRenderer
+    {
+        [Test]
+        public void NoFallback_ShowsDefaultInGeneratedShader()
```
Comment (joshua-davis) [inline TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/FeatureTests/FallbackTests.cs:11]: nit: I might rename these to be a bit clearer. "NoFallbackSet_DefaultFallbackIsUsed"

File: TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/FeatureTests/FallbackTests.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+            var container = new ShaderContainer();
+            var shaderCode = BlockTestRenderer.BuildSimpleSurfaceShader(container, "TestShader", Enumerable.Empty<Block>());
+
+            // LegacyTemplateProvider injects the shadergraph fallback to emulate ShaderGraph generator behavior
+            string expectedCode = "FallBack \"Hidden/Shader Graph/FallbackError\"";
+            Assert.AreNotEqual(-1, shaderCode.IndexOf(expectedCode), $"Missing expected code: {expectedCode}");
+        }
+
+        [Test]
+        public void Fallback_ShowsUpInGeneratedShader()
```
Comment (joshua-davis) [inline TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/FeatureTests/FallbackTests.cs:22]: nit: similar renaming for clarity.

File: TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/FeatureTests/IncludeTests.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+            var shaderCode = BlockTestRenderer.BuildSimpleSurfaceShader(container, "TestShader", blocks);
+            Debug.Log(shaderCode);
+
+            string expectedCode = "#include \"MyIncludeFile\"";
+            Assert.AreNotEqual(-1, shaderCode.IndexOf(expectedCode), $"Expected code missing: {expectedCode}");
+        }
+
+        [Test]
+        public void BlockFunctionInclude_ShowsUpInGeneratedShader()
```
Comment (joshua-davis) [inline TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/FeatureTests/IncludeTests.cs:33]: I'd try to improve the names a bit more to follow the "given when then" pattern. Something like
"BlockFunctionWithInclude_GeneratingShader_IncludeIsFound"

File: TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/FeatureTests/UsePassTests.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+namespace UnityEditor.ShaderFoundry.UnitTests
+{
+    class UsePassTests // : BlockTestRenderer
+    {
+        [Test]
+        public void NoUsePass_NotInGeneratedShader()
+        {
+            var container = new ShaderContainer();
+
+            SubShaderDescriptor ModifySubShader(SubShaderDescriptor subShader)
```
Comment (joshua-davis) [inline TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/FeatureTests/UsePassTests.cs:15]: nit: Is this even needed? Aren't these lists null by default?

File: TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/BlockLibrary/BlockBuilderUtilities.cs [hunk 13]
```diff
@@ -125,6 +125,35 @@ internal static Block.Builder CreateSimplePropertyBlockBuilder(ShaderContainer c
             return blockBuilder;
         }
 
+        internal static Block.Builder DefineSimpleBlock(ShaderContainer container, string blockName, ShaderType inputType, string inputName, ShaderType outputType, string outputName, string outputValue)
```
Comment (joshua-davis) [inline TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/BlockLibrary/BlockBuilderUtilities.cs:128]: nit: This function wasn't clear to me until I saw the use, a comment might be nice saying that it builds the input/output types from two fields.
Also maybe rename "outputValue" to "outputFieldExpression" or something.

File: com.unity.shadergraph/Editor/ShaderFoundry/Utilities/Extensions.cs [hunk 14]
```diff
@@ -191,11 +191,11 @@ internal static void MergeTypesAndFunctions(this Block.Builder builder, Block bl
             foreach (var item in block.ReferencedTypes)
                 builder.AddReferencedType(item);
             foreach (var item in block.Types)
-                builder.AddReferencedType(item);
+                builder.AddType(item);          // NOTE: parent block will be wrong...
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/ShaderFoundry/Utilities/Extensions.cs:194]: This is wrong. The new block doesn't own the type of the old block, it's a reference type. Why did you need to change this?

File: com.unity.shadergraph/Editor/ShaderFoundry/Utilities/Extensions.cs [hunk 15]
```diff
@@ -191,11 +191,11 @@ internal static void MergeTypesAndFunctions(this Block.Builder builder, Block bl
...
             foreach (var item in block.ReferencedFunctions)
                 builder.AddReferencedFunction(item);
             foreach (var item in block.Functions)
-                builder.AddReferencedFunction(item);
+                builder.AddFunction(item);      // NOTE: parent block will be wrong...
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/ShaderFoundry/Utilities/Extensions.cs:198]: Same as with type.

File: com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs [hunk 16]
```diff
@@ -8,20 +10,83 @@ namespace UnityEditor.ShaderFoundry
...
+            var dependencies = new List<ShaderDependency>();
+            foreach (var templateInstance in templateInstances)
+            {
+                if (!string.IsNullOrEmpty(templateInstance.Template.ShaderFallback))
+                {
+                    if (m_FallbackShader == null)
+                        m_FallbackShader = templateInstance.Template.ShaderFallback;
+                    else
+                    {
+                        if (m_FallbackShader != templateInstance.Template.ShaderFallback)
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs:44]: Won't this cause a problem if two independent template providers define a fallback making them incompatible?

File: com.unity.shadergraph/Editor/ShaderFoundry/Linker/LegacyTemplateLinker.cs [hunk 17]
```diff
@@ -112,6 +122,9 @@ void AddBlockInstanceProperties(BlockInstance blockInstance)
                 if (!blockInstance.IsValid)
                     return;
 
+                // this looks through block inputs, looking for any with a [Property] attribute, and assumes it is a property.
+                // this may need to be changed for VFX support, that replaces per-instance properties with override values
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/ShaderFoundry/Linker/LegacyTemplateLinker.cs:126]: nit: this will also need to change with link overrides.

File: com.unity.shadergraph/Editor/ShaderFoundry/Linker/LegacyTemplateLinker.cs [hunk 18]
```diff
@@ -563,6 +576,11 @@ void ProcessBlockInstance(BlockInstance blockInstance, VisitedRegistry visitedRe
                     shaderCommands.AddRange(block.Commands);
                     shaderDefines.AddRange(block.Defines);
                     shaderIncludes.AddRange(block.Includes);
+
+                    // TODO: there's probably a better way to gather includes from all functions -- like put this gathering in the visitor
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/ShaderFoundry/Linker/LegacyTemplateLinker.cs:580]: Yeah, ideally this would be done it the traversal that finds all used functions/types and declares them in the correct order, however this is fine as the order is mostly independent.

**Review by joshua-davis — CHANGES_REQUESTED** [review]

File: com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs (hunk 16 above)
Comment (cdxntchou) [reply com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderInstance.cs:44]: Yes.  ShaderLab only supports one fallback shader though, so a conflict is either an error, we choose one somehow, or we upgrade ShaderLab to do per-subshader fallbacks.  I chose error for now.  :)

File: com.unity.shadergraph/Editor/ShaderFoundry/Utilities/Extensions.cs (hunk 14 above)
Comment (joshua-davis) [reply com.unity.shadergraph/Editor/ShaderFoundry/Utilities/Extensions.cs:194]: Yeah, AddType is "owned" types, so they're in the block's namespace. Referenced types are any that you may use, so we can track them for knowing what types to emit.

File: TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/FeatureTests/IncludeTests.cs (hunk 11 above)
Comment (joshua-davis) [reply TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/FeatureTests/IncludeTests.cs:33]: https://internaldocs.unity.com/automation/TestFrameworks/NativeTests/BestPractices/
```
Name tests following the pattern: MethodOrFunctionName_ThisIsWhatIsExpected or MethodOrFunctionName_InputDescription_OutputOrBehaviourDescription or GivenSetup_WhenAction_ThenExpectedResult. Going with scenario-based names, is just as fine: CanAddFileWhileDatabaseIsLocked.
```

I think what you did is close to the first version, it's just not quite descriptive of what you're doing. I'll leave it up to you though.

**Review by joshua-davis — APPROVED** [review]
